enterprise/mqtt/src/main/java/io/crate/mqtt/protocol/MqttProcessor.java

```diff
@@ -0,0 +1,170 @@
+            }
+            NettyUtils.keepAlive(channel, keepAlive);
+
+            if (channel.pipeline().names().contains("idleStateHandler")) {
+                channel.pipeline().remove("idleStateHandler");
+            }
+            channel.pipeline().addFirst("idleStateHandler", new IdleStateHandler(0, 0, Math.round(keepAlive * 1.5f)));
+        }
+
+        channel.writeAndFlush(MqttMessageBuilders.connAck()
+                .returnCode(MqttConnectReturnCode.CONNECTION_ACCEPTED)
+                .sessionPresent(true)
+                .build())
+                .addListener(cf -> LOGGER.info("CONNECT sent CONNECTION_ACCEPTED"));
+    }
+
+    private ChannelFuture sendErrorResponse(Channel channel, MqttConnectReturnCode returnCode) {
+
+        return channel.writeAndFlush(MqttMessageBuilders.connAck()
+                .returnCode(returnCode)
+                .sessionPresent(false)
+                .build())
+                .addListener(ChannelFutureListener.CLOSE);
+    }
+
+    public void handlePublish(Channel channel, MqttPublishMessage msg) {
+        final MqttQoS qos = msg.fixedHeader().qosLevel();
+        String clientId = NettyUtils.clientID(channel);
+        switch (qos) {
+            case AT_LEAST_ONCE:
+                SendPubAckCallback ackCallback = new SendPubAckCallback(
+                        channel, msg.variableHeader().packetId(), msg.fixedHeader().isDup());
+                ingestService.doInsert(clientId, msg, ackCallback);
+                break;
+            default:
+                String message = String.format(Locale.ENGLISH, "QoS %s (%s) is not supported.",
+                        qos.value(), qos.name());
+                throw new UnsupportedOperationException(message);
+        }
+    }
+
+    public void handleDisconnect(Channel channel) throws InterruptedException {
+        channel.flush()
+                .close()
+                .addListener(cf -> LOGGER.debug("Client closed connection by sending DISCONNECT"));
+    }
+
+    public void handlePingReq(Channel channel) {
+        channel.writeAndFlush(MqttMessageBuilders.pingResp()
+            .qos(MqttQoS.AT_LEAST_ONCE)
+            .build())
+            .addListener(cf -> LOGGER.info("PINGRESP sent"));
+    }
+
+    private static class SendPubAckCallback implements BiConsumer<Object, Throwable> {
+
+        private final Channel channel;
+        private final Integer packetId;
+        private final boolean isDupFlag;
+
+        private SendPubAckCallback(Channel channel, Integer packetId, boolean isDupFlag) {
+            this.channel = channel;
+            this.packetId = packetId;
+            this.isDupFlag = isDupFlag;
+        }
+
+        @Override
+        public void accept(Object result, Throwable t) {
+            // If the DUP flag is set by the client a version conflict is ignored and a PubAck is returned.
+            if (t == null || (t.getClass() == VersionConflictEngineException.class && isDupFlag)) {
```

Comment:
Could we somehow avoid having to deal with internals like `VersonConflictEngineException` at this level?